packages/frontend/src/notebook/notebook_editor.tsx

```diff
@@ -249,60 +247,74 @@ export function NotebookEditor<T>(props: {
                 </div>
             </Show>
             <ul class="notebook-cells">
-                <For each={props.notebook.cells}>
-                    {(cell, i) => {
+                <For each={props.notebook.cellOrder}>
+                    {(cellId, i) => {
+                        const cell = props.notebook.cellContents[cellId];
+                        if (!cell) {
+                            throw `Failed to find contents for cell ${cellId}`;
+                        }
+
                         const isActive = () => activeCell() === i();
                         const cellActions: CellActions = {
                             activateAbove() {
                                 i() > 0 && setActiveCell(i() - 1);
                             },
                             activateBelow() {
-                                const n = props.notebook.cells.length;
+                                const n = props.notebook.cellOrder.length;
                                 i() < n - 1 && setActiveCell(i() + 1);
                             },
                             createAbove() {
                                 props.changeNotebook((nb) => {
-                                    nb.cells.splice(i(), 0, newStemCell());
+                                    const newCell = newStemCell();
+                                    nb.cellOrder.splice(i(), 0, newCell.id);
+                                    nb.cellContents[newCell.id] = newCell;
                                     setActiveCell(i());
                                 });
                             },
                             createBelow() {
+                                console.log("Creating below");
                                 props.changeNotebook((nb) => {
-                                    nb.cells.splice(i() + 1, 0, newStemCell());
+                                    const newCell = newStemCell();
+                                    nb.cellOrder.splice(i() + 1, 0, newCell.id);
+                                    nb.cellContents[newCell.id] = newCell;
+                                    // @ts-ignore
+                                    nb.cellStuff = { test: "huh" };
                                     setActiveCell(i() + 1);
                                 });
                             },
                             deleteBackward() {
                                 props.changeNotebook((nb) => {
-                                    nb.cells.splice(i(), 1);
+                                    const cellId = nb.cellOrder[i()];
+                                    // biome-ignore lint/style/noNonNullAssertion:
+                                    delete nb.cellContents[cellId!];
+                                    nb.cellOrder.splice(i(), 1);
```

Comment:
And perhaps also a helper function `deleteCell(notebook, i)`! Then you can actually check the invariant instead of having to put a non-null assertion.